src/logs/metadata_engine/jmespath.py

```diff
@@ -44,26 +44,96 @@ def _func_if(self, condition, if_true_expression, if_false_expression, node_scop
         else:
             return if_false_expression.visit(if_false_expression.expression, node_scope)
 
-    @functions.signature({"types": ['string']},
-                         {"types": ['array-string']})
+
+    @functions.signature({'types': ['string', 'null']}, {'types': ['string', 'null']})
+    def _func_starts_with(self, search, suffix):
+        if(search is None or suffix is None):
+            return False
+        return search.startswith(suffix)
+
+    @functions.signature({'types': ['string', 'null']},
+                         {"types": ['array']})
     def _func_format(self, pattern, values):
-        assert pattern.count("{}") == len(values)
+        return self.format(pattern, values)
 
-        output = pattern
+    def format(self, pattern, values):
+        if pattern.count("{}") != len(values):
+            return None
 
+        output = pattern
         for value in values:
-            output = output.replace("{}", value, 1)
-
+            if value is not None:
+                output = output.replace("{}", value, 1)
         return output
 
-    @functions.signature({'types': ['string']},
-                         {'types': ['string']})
+    @functions.signature({'types': ['string', 'null']},
+                         {"types": ['array']})
+    def _func_format_arn(self, pattern, values):
+        if None in values:
+            # any None value in input is treated as missing required data, no complete arn can be calculated
+            return None
+
+        return self.format(pattern, values)
+
+    @functions.signature({'types': ['string', 'null']},
+                         {'types': ['string', 'null']})
     def _func_dt_meid_md5(self, entity_type, hashing_input):
         return me_id.meid_md5(entity_type, hashing_input)
 
-    @functions.signature({'types': ['string']},
-                         {'types': ['string']})
+    @functions.signature({'types': ['string', 'null']},
+                         {'types': ['string', 'null']})
     def _func_dt_meid_murmurhash(self, entity_type, hashing_input):
         return me_id.meid_murmurhash(entity_type, hashing_input)
 
+    #same ids across all version of aws credentials - md5
+    @functions.signature({'types': ['string', 'null']})
+    def _func_dt_meid_ebs_volume(self, volumeId): return me_id.meid_md5("EBS_VOLUME", volumeId)
+    @functions.signature({'types': ['string', 'null']})
+    def _func_dt_meid_ec2_instance(self, instanceId): return me_id.meid_md5("EC2_INSTANCE", instanceId)
+    @functions.signature({'types': ['string', 'null']},{'types': ['string', 'null']},{'types': ['string', 'null']})
+    def _func_dt_meid_lambda_function(self, functionName, region, accountId): return me_id.meid_md5("AWS_LAMBDA_FUNCTION", functionName, region, "_", accountId)
```

Comment:
I feel you should prepare hashing input as a one string here so
region, "_", accountId -> region+"_"+accountId 
And not do any concatenations in hashing methods. 
But this is minor, feel free to ignore

Reply:
Agreed that it's minor but I also favor the `format("{}{}_{}", functionName, region, accountId) notation. It also clearly separates the first parameter (`entity_type`) from `hashing_input`. Here those two parts are "mixed together"